### PR TITLE
S3 Key sensor deferrable

### PR DESCRIPTION
**Pull request by syedahsn**
The PR allows the S3KeySensor to be run in deferrable mode. I refactored the existing S3KeySensor to pull out the API calls, and leave a common `process_files` method that is used in both the deferrable and non-deferrable case. This reduces a lot of the code duplication. 

Some duplication is unavoidable - like writing the `head_object_async` and `get_file_metadata_async`, but where possible, I tried code duplication to a minimum. 

The unit tests for the S3KeySensorTrigger follow the same pattern of testing as the S3KeySensor.

@sunank200 @ephraimbuddy @pankajkoti 

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by pankajkoti** [comment]
@syedahsn perhaps you meant to tag @pankajastro :)

**Comment by sunank200** [comment]
There is already a PR I have created here for the same: https://github.com/apache/airflow/pull/31018

**Comment by ephraimbuddy** [comment]
@syedahsn, you probably missed my comment in the other PR, See https://github.com/apache/airflow/pull/31018#issuecomment-1567947772. Since duplication is the issue, can you do this when the other PR is merged? The other PR also added some dags which we can run in system tests when deferrable on system tests is supported

File: airflow/providers/amazon/aws/sensors/s3.py [hunk 1]
```diff
@@ -93,54 +95,102 @@ def __init__(
...
+            self.bucket_keys = self.bucket_key
+        wildcard_keys = []
+        objs = []
+        bucket_key_names = []
+        for i in range(len(self.bucket_keys)):
+            bucket_key_names.append(
+                S3Hook.get_s3_bucket_key(self.bucket_name, self.bucket_keys[i], "bucket_name", "bucket_key")
+            )
+            bucket_name = bucket_key_names[i][0]
+            key = bucket_key_names[i][1]
```
Comment (uranusjr) [inline airflow/providers/amazon/aws/sensors/s3.py:130]: This can use a rewrite with `enumerate`

File: airflow/providers/amazon/aws/sensors/s3.py [hunk 2]
```diff
@@ -93,54 +95,102 @@ def __init__(
...
+        self.log.info("Inside execute complete")
+        if event["status"] != "success":
+            raise AirflowException(f"Error: {event}")
+        else:
+            results = []
+            self.log.info("Success: %s", event)
+            if self.check_fn is not None:
+                for files in event["files_list"]:
+                    results.append(self.check_fn(files))
+                return all(results)
```
Comment (uranusjr) [inline airflow/providers/amazon/aws/sensors/s3.py:158]: Why build the entire `results` list and then call `all()`? Instead you can just do

```python
for f in event["files_list"]:
    if not self.check_fn(f):
        return False
return True
```

or

```python
return all(self.check_fn(f) for f in event["files_list"])
```

File: airflow/providers/amazon/aws/sensors/s3.py [hunk 3]
```diff
@@ -93,54 +95,102 @@ def __init__(
...
+    files_list = []
+    for i in range(len(bucket_keys)):
+        key = bucket_key_names[i][1]
+        if wildcard_match:
+            key_matches = [k for k in wildcard_keys[i] if fnmatch.fnmatch(k["Key"], key)]
+            if len(key_matches) == 0:
+                results.append(False)
+                continue
+            # Reduce the set of metadata to size only
+            files_list.append(list(map(lambda f: {"Size": f["Size"]}, key_matches)))
```
Comment (uranusjr) [inline airflow/providers/amazon/aws/sensors/s3.py:176]: This list-building code can be improved by using iterators

File: airflow/providers/amazon/aws/triggers/s3.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    async def poke(self):
+        if isinstance(self.bucket_key, str):
+            self.bucket_keys = [self.bucket_key]
+        else:
+            self.bucket_keys = self.bucket_key
+
+        wildcard_keys = []
+        obj = []
+        bucket_key_names = []
+        for i in range(len(self.bucket_keys)):
```
Comment (uranusjr) [inline airflow/providers/amazon/aws/triggers/s3.py:76]: This loop looks very much like `poke` in `S3KeySensor`. Can they share one single implementation?

**Comment by syedahsn** [comment]
> @syedahsn, you probably missed my comment in the other PR, See [#31018 (comment)](https://github.com/apache/airflow/pull/31018#issuecomment-1567947772). Since duplication is the issue, can you do this when the other PR is merged? The other PR also added some dags which we can run in system tests when deferrable on system tests is supported

I saw your comment, but I decided to open this PR to get your thoughts (as well as the community's) on the approach I took for this sensor. As I mentioned in the initial PR for this sensor, one concern is the amount of code that is being duplicated for the async case. I think that keeping code duplication to a minimum is very important because it will lead to a code base that is easier to maintain, and less error-prone. 

I'm willing to wait for the merge of https://github.com/apache/airflow/pull/31018 before addressing the code duplication, but because of the differences in our approaches, it would mean that I would end up removing a lot of the code introduced in https://github.com/apache/airflow/pull/31018. 

I'll leave it to you to decide whether we should merge the inital PR and I address the code duplication in a follow-up PR or whether we collaborate now to come up with a suitable solution.

**Pull request closed without merging**